File: lintcode/python/0059_3sum_closest.py
#!/usr/bin/python -t

# two pointers

class Solution:
    """
    @param numbers: Give an array numbers of n integer
    @param target: An integer
    @return: return the sum of the three integers, the sum closest target.
    """
    def threeSumClosest(self, numbers, target):
        # write your code here
        numbers.sort()
        
        n = len(numbers)
        diff = sys.maxint
        ret = 0
        
        for i in range(n-2):
            l = i+1
            r = n-1
            while l < r:
                val = numbers[l]+numbers[r]+numbers[i]
                if abs(val-target) < diff:
                    diff = abs(val-target)
                    ret = val
                if val == target:
                    return ret
                elif val > target:
                    r -= 1
                else:
                    l += 1

                        
        return ret
        

File: lintcode/golang/0059_3sum_closest.go
/**
 * @param numbers: Give an array numbers of n integer
 * @param target: An integer
 * @return: return the sum of the three integers, the sum closest target.
 */
 
import "sort"

func threeSumClosest (numbers []int, target int) int {
    // write your code here
    if len(numbers) == 0 {
        return 0
    }
    
    sort.Ints(numbers)
    var ret int
    var sum int
    minDiff := target
    
    for index := range numbers {
        l := index+1
        r := len(numbers)-1

        for l < r {
            sum = numbers[index] + numbers[l] + numbers[r]
            
            if sum == target {
                return sum
            }
            
            if ret == 0 || abs(sum-target) < minDiff {
                minDiff = abs(sum-target)
                ret = sum
            }
            
            if sum > target {
                r--
            } else {
                l++
            }
        }
    }
    
    return ret
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}

